docs/guides/streaming.md

```diff
@@ -0,0 +1,349 @@
+  const data = useLoaderData() as LoaderData;
+
+  return (
+    <main>
+      <h1>Let's locate your package</h1>
+      <Deferred
+        value={data.packageLocation}
+        fallback={<p>Loading package location...</p>}
+        errorBoundary={
+          <p>Error loading package location!</p>
+        }
+      >
+        {(packageLocation) => (
+          <p>
+            Your package is at {packageLocation.latitude}{" "}
+            lat and {packageLocation.longitude} long.
+          </p>
+        )}
+      </Deferred>
+    </main>
+  );
+}
+```
+
+<details>
+  <summary>Alternatively, you can use the `useDeferredValue` hook:</summary>
+
+If you're not jazzed about bringing back render props, you can use a hook, but you'll have to break things out into another component:
+
+```tsx lines=[14,21]
+export default function PackageRoute() {
+  const data = useLoaderData() as LoaderData;
+
+  return (
+    <main>
+      <h1>Let's locate your package</h1>
+      <Deferred
+        value={data.packageLocation}
+        fallback={<p>Loading package location...</p>}
+        errorBoundary={
+          <p>Error loading package location!</p>
+        }
+      >
+        <PackageLocation />
+      </Deferred>
+    </main>
+  );
+}
+
+function PackageLocation() {
+  const packageLocation = useDeferredValue();
+  return (
+    <p>
+      Your package is at {packageLocation.latitude} lat and{" "}
+      {packageLocation.longitude} long.
+    </p>
+  );
+}
+```
+
+</details>
+
+## Evaluating the solution
+
+So rather than waiting for the whole `document -> JavaScript -> hydrate -> request` cycle, with streaming we start the request for the slow data as soon as the document request comes in. This can significantly speed up the user experience.
+
+Additionally, the API that Remix exposes for this is extremely ergonomic. You can literally switch between whether something is going to be derferred based on whether you include the `await` keyword:
+
+```tsx
+return deferred<LoaderData>({
+  // not deferred:
+  packageLocation: await packageLocationPromise,
+  // deferred:
+  packageLocation: packageLocationPromise,
+});
+```
+
+Because of this, you can A/B test deferring, or even determine whether to defer based on the user or data being requested:
+
+```tsx
+export const loader: LoaderFunction = ({
+  request,
+  params,
+}) => {
+  const packageLocationPromise = getPackageLocation(
+    params.packageId
+  );
+  const shouldDefer = await shouldDeferPackageLocation(
+    request,
+    params.packageId
+  );
+
+  return deferred<LoaderData>({
+    packageLocation: shouldDefer
+      ? packageLocationPromise
+      : await packageLocationPromise,
+  });
+};
+```
+
+That `shouldDeferPackageLocation` could be implemented to check the user making the request, whether the package location data is in a cache, the status of an A/B test, or whatever else you want. This is pretty sweet 🍭
+
+## When to not use streaming
+
+TODO: Talk about no streaming on action reloads.
```

Comment:
@jacob-ebey, can you help me come up with an example of skipping deferring on revalidations?

Reply:
Jacob should confirm but I believe `deferred` should be automatically "disabled" on revalidations as we just await all the promises before resolving the `loader` if it's not an initial request